crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+        if self.block_indices.contains_block_hash(block.hash()) {
+            // block is known return that is inserted
+            return Ok(true)
+        }
+
+        // check if block is part of canonical chain
+        if self.block_indices.canonical_hash(&block.number) == Some(block.hash()) {
+            // block is part of canonical chain
+            return Ok(true)
+        }
+
+        // check if block parent can be found in Tree
+        if let Some(parent_chain) = self.block_indices.get_blocks_chain_id(&block.parent_hash) {
+            self.fork_side_chain(block.clone(), parent_chain)?;
+            //self.db.tx_mut()?.put::<tables::PendingBlocks>(block.hash(), block.unseal())?;
+            return Ok(true)
+        }
+
+        // if not found, check if it can be found inside canonical chain.
+        if Some(block.parent_hash) == self.block_indices.canonical_hash(&(block.number - 1)) {
+            // create new chain that points to that block
+            self.fork_canonical_chain(block.clone())?;
+            //self.db.tx_mut()?.put::<tables::PendingBlocks>(block.hash(), block.unseal())?;
+            return Ok(true)
+        }
+        // NOTE: Block doesn't have a parent, and if we receive this block in `make_canonical`
+        // function this could be a trigger to initiate p2p syncing, as we are missing the
+        // parent.
+        Ok(false)
+    }
+
+    /// Do finalization of blocks. Remove them from tree
+    pub fn finalize_block(&mut self, finalized_block: BlockNumber) {
+        let mut remove_chains = self.block_indices.finalize_canonical_blocks(finalized_block);
+
+        while let Some(chain_id) = remove_chains.pop_first() {
+            if let Some(chain) = self.chains.remove(&chain_id) {
+                remove_chains.extend(self.block_indices.remove_chain(&chain));
+            }
+        }
+    }
+
+    /// Update canonical hashes. Reads last N canonical blocks from database and update all indices.
+    pub fn update_canonical_hashes(
+        &mut self,
+        last_finalized_block: BlockNumber,
+    ) -> Result<(), Error> {
+        self.finalize_block(last_finalized_block);
+
+        let num_of_canonical_hashes =
+            self.finalization_window + self.block_indices.num_of_additional_canonical_block_hashes;
+
+        let last_canonical_hashes = self
+            .externals
+            .db
+            .tx()?
+            .cursor_read::<tables::CanonicalHeaders>()?
+            .walk_back(None)?
+            .take(num_of_canonical_hashes as usize)
+            .collect::<Result<BTreeMap<BlockNumber, BlockHash>, _>>()?;
+
+        let mut remove_chains = self.block_indices.update_block_hashes(last_canonical_hashes);
+
+        // remove all chains that got discarded
+        while let Some(chain_id) = remove_chains.first() {
+            if let Some(chain) = self.chains.remove(chain_id) {
+                remove_chains.extend(self.block_indices.remove_chain(&chain));
+            }
+        }
```

Comment:
It is a little bit different.

in `finalize_block` we are working with block numbers, and any block/chain that depends on block number that is less then finalize block will be removed.

in `update_block_hashes` (Think about pipeline updating canonical blocks) we are working on canonical chain and and replacing block hashes if needed or adding new one if updated chain is longer. and if canonical chain got replaced any block or chain that depends on it will be removed.